File: backend/app/services/recipe.py
"""OpenAI recipe generation service — ingredients to structured recipes."""

import json
import logging

from openai import AsyncOpenAI

from app.config import get_settings
from app.models.recipe import InstructionStep, RecipeDetail
from app.prompts.recipe import RECIPE_RESPONSE_FORMAT, RECIPE_SYSTEM_PROMPT

logger = logging.getLogger(__name__)

_client: AsyncOpenAI | None = None


def _get_client() -> AsyncOpenAI:
    global _client
    if _client is None:
        _client = AsyncOpenAI(api_key=get_settings().openai_api_key)
    return _client


async def generate_recipes(
    ingredients: list[str],
    num_recipes: int | None = None,
) -> list[RecipeDetail]:
    """Generate recipes from ingredients using OpenAI with structured output.

    Uses response_format with json_schema for guaranteed valid JSON.
    """
    settings = get_settings()
    client = _get_client()
    count = num_recipes or settings.recipe_results_limit

    user_prompt = (
        f"Ingredients available: {', '.join(ingredients)}\n\n"
        f"Generate exactly {count} different recipes using these ingredients. "
        "Vary the cuisine styles and cooking methods."
    )

    response = await client.chat.completions.create(
        model=settings.model_name,
        messages=[
            {"role": "system", "content": RECIPE_SYSTEM_PROMPT},
            {"role": "user", "content": user_prompt},
        ],
        response_format=RECIPE_RESPONSE_FORMAT,
        max_completion_tokens=8192,
        reasoning_effort="low",
    )

    raw_text = response.choices[0].message.content
    logger.info("OpenAI recipe response length: %d chars", len(raw_text))

    parsed = json.loads(raw_text)
    recipe_items = parsed.get("recipes", parsed) if isinstance(parsed, dict) else parsed

    recipes: list[RecipeDetail] = []
    for item in recipe_items:
        steps = [
            InstructionStep(number=s["number"], step=s["step"])
            for s in item.get("steps", [])
        ]
        recipes.append(
            RecipeDetail(
                title=item["title"],
                ready_in_minutes=item.get("ready_in_minutes", 0),
                servings=item.get("servings", 0),
                summary=item.get("summary", ""),
                ingredients_used=item.get("ingredients_used", []),
                ingredients_extra=item.get("ingredients_extra", []),
                steps=steps,
            )
        )

    return recipes

File: frontend/kitchengpt/lib/models/recipe.dart
/// Mirrors the backend RecipeDetail model (Gemini-generated).
class InstructionStep {
  final int number;
  final String step;

  InstructionStep({required this.number, required this.step});

  factory InstructionStep.fromJson(Map<String, dynamic> json) {
    return InstructionStep(
      number: json['number'] as int,
      step: json['step'] as String,
    );
  }
}

class Recipe {
  final String title;
  final int readyInMinutes;
  final int servings;
  final String summary;
  final List<String> ingredientsUsed;
  final List<String> ingredientsExtra;
  final List<InstructionStep> steps;

  Recipe({
    required this.title,
    required this.readyInMinutes,
    required this.servings,
    required this.summary,
    required this.ingredientsUsed,
    required this.ingredientsExtra,
    required this.steps,
  });

  factory Recipe.fromJson(Map<String, dynamic> json) {
    return Recipe(
      title: json['title'] as String,
      readyInMinutes: json['ready_in_minutes'] as int? ?? 0,
      servings: json['servings'] as int? ?? 0,
      summary: json['summary'] as String? ?? '',
      ingredientsUsed: List<String>.from(json['ingredients_used'] as List? ?? []),
      ingredientsExtra:
          List<String>.from(json['ingredients_extra'] as List? ?? []),
      steps: (json['steps'] as List?)
              ?.map((s) => InstructionStep.fromJson(s as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }
}
